packages/core/crates/core-strategy/src/generic.rs

```diff
@@ -0,0 +1,220 @@
+use utils_types::primitives::Balance;
+
+/// Basic strategy trait that all strategies must implement.
+/// Strategies make decisions to automatically open/close certain channels.
+/// The decision is done by the `tick` method based on the current node's balance, current view
+/// of the network (all peer ids, currently opened outgoing channels) and quality estimations
+/// of connections to the other peers in the network.
+/// Other additional parameters affecting this decision could be members of implementors
+/// of this trait.
+pub trait ChannelStrategy {
+    /// Human readable name of the strategy
+    const NAME: &'static str;
+
+    /// Human readable name of the strategy
+    fn name(&self) -> &str {
+        Self::NAME
+    }
+
+    /// Performs the strategy tick - deciding which new channels should be opened and
+    /// which existing channels should be closed.
+    fn tick<Q>(
+        &self,
+        balance: Balance,
+        peer_ids: impl Iterator<Item = String>,
+        outgoing_channels: Vec<OutgoingChannelStatus>,
+        quality_of_peer: Q,
+    ) -> StrategyTickResult
+    where
+        Q: Fn(&str) -> Option<f64>;
+
+    /// Indicates if according to this strategy, a commitment should be made for the given channel.
+    fn should_commit_to_channel(&self, _channel: &OutgoingChannelStatus) -> bool {
+        true
+    }
+}
+
+/// Represents a request to open a channel with a stake.
+pub struct OutgoingChannelStatus {
+    pub peer_id: String,
+    pub stake: Balance,
+}
+
+#[cfg(feature = "wasm")]
+impl From<&wasm::OutgoingChannelStatus> for OutgoingChannelStatus {
+    fn from(x: &wasm::OutgoingChannelStatus) -> Self {
+        OutgoingChannelStatus {
+            peer_id: x.peer_id.clone(),
+            stake: Balance::from_str(x.stake_str.as_str()).unwrap(),
+        }
+    }
+}
+
+/// A decision made by a strategy on each tick,
+/// represents which channels should be closed and which should be opened.
+/// Also indicates a number of maximum channels this strategy can open given the current network size.
+/// Note that the number changes as the network size changes.
+pub struct StrategyTickResult {
+    max_auto_channels: usize,
```

Comment:
The aim is to have the struct always immutable. If you make the field accessible via `pub`, it can be changed if the the whole struct is instantiated with `mut`.